cache/private/census_topic.go

```diff
@@ -0,0 +1,139 @@
+package private
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+
+	"github.com/ONSdigital/dp-frontend-homepage-controller/cache"
+	"github.com/ONSdigital/dp-topic-api/models"
+	topicCli "github.com/ONSdigital/dp-topic-api/sdk"
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+// UpdateCensusTopic is a function to update the census topic cache in publishing (private) mode.
+// This function talks to the dp-topic-api via its private endpoints to retrieve the census topic and its subtopic ids
+// The data returned by the dp-topic-api is of type *models.PrivateSubtopics which is then transformed in this function for the controller
+// If an error has occurred, this is captured in log.Error and then an empty census topic is returned
+func UpdateCensusTopic(ctx context.Context, serviceAuthToken string, topicClient topicCli.Clienter) func() *cache.Topic {
+	return func() *cache.Topic {
+		// get root topics from dp-topic-api
+		rootTopics, err := topicClient.GetRootTopicsPrivate(ctx, topicCli.Headers{ServiceAuthToken: serviceAuthToken})
+		if err != nil {
+			logData := log.Data{
+				"req_headers": topicCli.Headers{},
+			}
+			log.Error(ctx, "failed to get root topics from topic-api", err, logData)
+			return cache.GetEmptyCensusTopic()
+		}
+
+		// deference root topics items to allow ranging through them
+		var rootTopicItems []models.TopicResponse
+		if rootTopics.PrivateItems != nil {
+			rootTopicItems = *rootTopics.PrivateItems
+		} else {
+			err := errors.New("root topic public items is nil")
+			log.Error(ctx, "failed to deference root topics items pointer", err)
+			return cache.GetEmptyCensusTopic()
+		}
+
+		var censusTopicCache *cache.Topic
+
+		// go through each root topic, find census topic and gets its data for caching which includes subtopic ids
+		for i := range rootTopicItems {
```

Comment:
Do we need to call root topic to get a list of root topics and the loop through topics to find census id, we already know the id so we can use the `cache.CensusTopicID` to call it's subtopics?